### ISSUE TITLE
Mixin for lattice

### ISSUE THREAD

**Issue by hivert** [math-comp/math-comp#762]
Currently the mixin for lattice contains a lot of obligations (this is the algebraic point of view on lattices):
```
Record mixin_of (T0 : Type) (b : POrder.class_of T0)
                (T := POrder.Pack tt b) := Mixin {
  meet : T -> T -> T;
  join : T -> T -> T;
  _ : commutative meet;
  _ : commutative join;
  _ : associative meet;
  _ : associative join;
  _ : forall y x, meet x (join x y) = x;
  _ : forall y x, join x (meet x y) = x;
  _ : forall x y, (x <= y) = (meet x y == x);
}.
```
We only need to require that meet is idempotent (ie: meet x x = x), to prove that `<=` is a partial order (see below). So I feel that this mixin is too much to ask.

If we already know that `<=` is a partial order, then a simpler mixin could be (this is the order theory point of view):
```
Record of_ := Build {
  meet : T -> T -> T;
  join : T -> T -> T;
  meetP : forall x y z, (x <= meet y z) = (x <= y) && (x <= z);
  joinP : forall x y z, (join x y <= z) = (x <= z) && (y <= z);
}.
```

I'm not sure if we want to use the order-theory mixin only at creation or to change the internal structure of the mixin to the above one, using coercion to allow creation from different mixin for backward compatibility. 

If we go to the input only solution, I've code at https://github.com/hivert/Coq-Combi/blob/master/theories/SSRcomplements/lattice.v which create a latticeMixin from a mixin above. I'm ok to submit a PR.

#### Here is a proof of my claim:
```
Variables (T : eqType) (meet : T -> T -> T).
Hypothesis (meetI : forall x, meet x x = x)
           (meetC : commutative meet)
           (meetA : associative meet).

Definition le x y := (meet x y == x).

Lemma le_refl x : le x x.
Proof. by rewrite /le meetI. Qed.
Lemma le_trans : transitive le.
Proof.
move=> x y z; rewrite /le => /eqP myx /eqP mxz.
by rewrite -myx -meetA mxz.
Qed.
Lemma le_anti : antisymmetric le.
Proof.
move=> x y; rewrite /le => /andP [/eqP mxy].
by rewrite meetC mxy => /eqP.
Qed.
```

**Comment by CohenCyril** [comment]
Hi @hivert thanks for this remark. It does not matter much what is the internal representation as long as the names, theory stay the same : additional mixins can fill the gap. I would be happy with the a minima change you suggest for mathcomp 1.x.x. We could replace the internal representation in the HB branch later on.

**Comment by hivert** [comment]
Ok ! I'm putting together the a minima pull request.

**Comment by hivert** [comment]
Should I close the issue, or leave it open as a remainder if you want to change the internal representation during the passage to HB ? By the way, I'd be happy to help doing that, it could be a way for me to learn HB.

**Comment by chdoc** [comment]
If your minimal PR has been merged, I think this issue should be be closed.

**Comment by gares** [comment]
You are welcome to join, the porting to HB is at a good state, but not finished yet. Maybe we should come back to it with a plan after the summer, to ease the contribution of others @CohenCyril

**Comment by affeldt-aist** [comment]
This seems to have been resolved by PR #763, please reopen if not.
